### DISCUSSION TITLE
Intellisense (typedefs) for your eleventy configuration file

### DISCUSSION THREAD

**Discussion by binyamin** [11ty/eleventy#2089] [Show and tell]
I just thought this would be useful for people. I'd love to see a short blog post on this, if someone has time.

```js
/**
 * @typedef {import("@11ty/eleventy/src/UserConfig")} EleventyConfig
 * @typedef {ReturnType<import("@11ty/eleventy/src/defaultConfig")>} EleventyReturnValue
 * 
 * @type {(eleventyConfig: EleventyConfig) => EleventyReturnValue}
 */
module.exports = function(eleventyConfig) {
    // write some code...
}
```

**Edit:** _Shorter version_
```js
/**
 *  @param {import("@11ty/eleventy/src/UserConfig")} eleventyConfig
 *  @returns {ReturnType<import("@11ty/eleventy/src/defaultConfig")>}
 */
module.exports = function(eleventyConfig) {
    // some code...
}
```

**Comment by cfjedimaster** [comment]
Is this right? `@type {(eleventyConfig: EleventyConfig)) => EleventyReturnValue}` Looks like an extra )

**Reply by binyamin** [reply]
The way that code is working, it defines the function with `@type {(paramName: ParamType) => ReturnValueType}`. the two `@typedef {...}` lines are utilities, so that it's not too messy.

The truth is that the following method would be shorter _and_ clearer. ~~Not tested though.~~ (_Edit_: tested)
```js
/**
 *  @param {import("@11ty/eleventy/src/UserConfig")} eleventyConfig
 *  @returns {ReturnType<import("@11ty/eleventy/src/defaultConfig")>}
 */
module.exports = function(eleventyConfig) {
    // some code...
}
```

**Reply by pdehaan** [reply]
I've used the `@param {import("@11ty/eleventy/src/UserConfig")} eleventyConfig` version before, but the `defaultConfig` bit is new to me and not something I would have thought to try and solve, so thanks for that!

**Reply by wommy** [reply]
so is there supposed to be a second `)` or is that a typo?

in [@binyamin 's example](https://github.com/11ty/eleventy/discussions/2089#discussioncomment-1647706) he only uses one `)`

seems like a mistake

**Reply by binyamin** [reply]
@wommy fixed, thanks!

**Reply by wommy** [reply]
@binyamin you mentioned earlier that YMMV with `@returns`

is that still true? or can we rely on it

cc: https://github.com/BenDMyers/serverless-contrast-starter/issues/1

**Reply by binyamin** [reply]
@wommy Yeah, it does work. I just edited the comment. Thanks.

**Comment by pdehaan** [comment]
> I'd love to see a short blog post on this, if someone has time.

I think we should file this as an official documentation request in https://github.com/11ty/11ty-website. Not sure where the final docs would love (Getting Started? somewhere else?) but having code hinting documented would be a huge improvement to my workflows and help me from mistyping `eleventyConfig.setDataDeepMerge(true)` wrong about 2/3rds of the time (although that will change in v1.0.0 anyways since that's the new default, if I remember correctly).

**Comment by wommy** [comment]
just found this: https://github.com/11ty/eleventy/issues/577#issuecomment-1126862063

**Comment by zachleat** [comment]
For future folks this is on the docs here: https://www.11ty.dev/docs/config/#type-definitions
